### PR TITLE
AsyncMessage handler

### PR DESCRIPTION
**Pull request by sirkonst**

### CODE REVIEW

**Comment by kozzztik** [comment]
I dont really sure, that you must force loop as handler kwarg. User may create on object with loop parameter and provide object method where loop can be accessible through self variable.

**Comment by sirkonst** [comment]
In this case, problems can arise when instance of AsyncMessage class created in one thread and smtpd server runn in other. And this threads has different loops.  It's real case if use aiosmtpd.controller.Controller like http://aiosmtpd.readthedocs.io/en/latest/aiosmtpd/docs/controller.html

**Comment by warsaw** [comment]
I think you're right about that.  Given the current structure of the code, the best way to handle that would be to pass the `AsyncMessage` *class* in as the handler, and override the `factory()` function in a subclass of the `Controller` to instantiate it.  E.g. (untested):

```python

class AsyncMessageController(Controller):
    def factory(self):
        return SMTP(self.handler(self.loop))

my_controller = AsyncMessageController(AsyncMessage)
```

File: aiosmtpd/handlers.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 your own handling of messages.  Implement only the methods you care about.
 """
 
-
+import asyncio
```
Comment (warsaw) [inline aiosmtpd/handlers.py]: This will fail `tox -e qa`.  Please rebase/merge master and you'll see the failures.  We have a few custom flake8 rules to enforce import order.  E.g.:

```
aiosmtpd/handlers.py:10:1: B403 Shorter non-from import follows longer
aiosmtpd/tests/test_handlers.py:2:1: B403 Shorter non-from import follows longer
```

File: aiosmtpd/handlers.py [hunk 2]
```diff
@@ -155,6 +155,29 @@ def handle_message(self, message):
 
 
 @public
+class AsyncMessage(Message):
+
```
Comment (warsaw) [inline aiosmtpd/handlers.py]: I'd like to see a docstring here, to briefly describe the uses of this class.  E.g. Why use this over the base class?

File: aiosmtpd/handlers.py [hunk 3]
```diff
@@ -155,6 +155,29 @@ def handle_message(self, message):
...
+        if isinstance(data, bytes):
+            message = message_from_bytes(data, self.message_class)
+        else:
+            assert isinstance(data, str), (
+              'Expected str or bytes, got {}'.format(type(data)))
+            message = message_from_string(data, self.message_class)
+        message['X-Peer'] = str(peer)
+        message['X-MailFrom'] = mailfrom
+        message['X-RcptTos'] = COMMASPACE.join(rcpttos)
+        yield from self.handle_message(message, loop=loop)
```
Comment (warsaw) [inline aiosmtpd/handlers.py]: Nearly all of this is shared with the `Message.process_message()` base class method.  Is there a good way to refactor the code so that it doesn't have to be duplicated?  That would make it much easier if we ever have to change something (e.g. add another default `X-` header.

File: aiosmtpd/smtp.py [hunk 4]
```diff
@@ -439,7 +439,13 @@ def smtp_DATA(self, arg):
                 'mail_options': self.mail_options,
                 'rcpt_options': self.rcpt_options,
             }
-        status = self.event_handler.process_message(*args, **kwargs)
+        if asyncio.iscoroutinefunction(self.event_handler.process_message):
+            kwargs['loop'] = self.loop
+            status = yield from self.event_handler.process_message(
+                *args, **kwargs
+            )
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: Our coding style would prefer:

```
            status = yield from self.event_handler.process_message(
                *args, **kwargs)
```

File: aiosmtpd/tests/test_handlers.py [hunk 5]
```diff
@@ -1,9 +1,11 @@
+import asyncio
 import os
 import sys
 import unittest
 
 from aiosmtpd.controller import Controller
-from aiosmtpd.handlers import Debugging, Mailbox, Message, Proxy, Sink
+from aiosmtpd.handlers import Debugging, Mailbox, Message, Proxy, Sink, \
+    AsyncMessage
```
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: Please use parentheses instead of backslashes.   Also, the imported names must be alphabetically sorted.  E.g.:

```
from aiosmtpd.handlers import (
    AsyncMessage, Debugging, Mailbox, Message, Proxy, Sink)
```

File: aiosmtpd/tests/test_handlers.py [hunk 6]
```diff
@@ -186,6 +188,65 @@ def test_message_decoded(self):
...
+class TestAsyncMessage(unittest.TestCase):
+    def setUp(self):
+        self.handled_message = None
+
+        class MessageHandler(AsyncMessage):
+            @asyncio.coroutine
+            def handle_message(handler_self, message, loop):
+                self.handled_message = message
+
+        self.handler = MessageHandler()
```
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: I'm not a big fan of vertical whitespace for separation.  I'll suggest adding comments instead, or restructuring the test to make the subclass a module global, or just collapsing out the whitespace (it's not generally too hard to see the difference).  But I'm open to thoughts here.

Also, It seems like you could create the controller, start it, and arrange for it to be stopped in an `addCleanup()` all in the `setUp()` method.  That would reduce some duplication in the test methods.

File: aiosmtpd/tests/test_handlers.py [hunk 7]
```diff
@@ -186,6 +188,65 @@ def test_message_decoded(self):
...
+
+    def test_message(self):
+        # In this test, the message data comes in as bytes.
+        controller = Controller(self.handler)
+        controller.start()
+        self.addCleanup(controller.stop)
+
```
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: Here's another blank line that's not needed.

**Comment by warsaw** [comment]
Thanks for the contribution!  Other than the review comments, I would like to ask if you could add some documentation for the `AsyncMessage` handler.  I wish we had better documentation for the base `Message` class too.  Would you consider adding a .rst file for documenting the `AsyncMessage` class and why you'd choose it over say the base `Message` class?

File: aiosmtpd/handlers.py (hunk 1 above)
Comment (sirkonst) [inline aiosmtpd/handlers.py]: This warning about line 10 (import sys) how I see, it is not my code.

File: aiosmtpd/tests/test_handlers.py (hunk 6 above)
Comment (sirkonst) [inline aiosmtpd/tests/test_handlers.py]: I just took code from TestMessage and rewrote for AsyncMessage. Your points is right but I have not goal refact all other tests, I've just want to add new a feature :-)

**Comment by sirkonst** [comment]
I rebased to master and answered for review comments.

**Comment by sirkonst** [comment]
Ping, please review.

File: aiosmtpd/handlers.py [hunk 8]
```diff
@@ -148,13 +148,34 @@ def process_message(self, peer, mailfrom, rcpttos, data, **kws):
         message['X-Peer'] = str(peer)
         message['X-MailFrom'] = mailfrom
         message['X-RcptTos'] = COMMASPACE.join(rcpttos)
+
+        return message
+
+    def process_message(self, peer, mailfrom, rcpttos, data, **kws):
+        message = self.prepare_message(peer, mailfrom, rcpttos, data, **kws)
```
Comment (warsaw) [inline aiosmtpd/handlers.py]: I'm curious as to why you split this out into `prepare_message()`.  I don't necessarily think it's a bad thing to do, just curious as to your thinking here.

File: aiosmtpd/handlers.py (hunk 1 above)
Comment (warsaw) [inline aiosmtpd/handlers.py]: I don't see these happening in master though.

File: aiosmtpd/smtp.py [hunk 9]
```diff
@@ -438,7 +438,12 @@ def smtp_DATA(self, arg):
                 'mail_options': self.mail_options,
                 'rcpt_options': self.rcpt_options,
             }
-        status = self.event_handler.process_message(*args, **kwargs)
+        if asyncio.iscoroutinefunction(self.event_handler.process_message):
+            kwargs['loop'] = self.loop
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: I'm curious as to why you do this only for `iscoroutinefunction`s?  What's the harm in doing it unconditionally, or checking something else (e.g. that `self.loop is not None` or some such)?

File: aiosmtpd/smtp.py [hunk 10]
```diff
@@ -438,7 +438,12 @@ def smtp_DATA(self, arg):
...
+            status = yield from self.event_handler.process_message(
+                *args, **kwargs)
+        else:
+            status = self.event_handler.process_message(*args, **kwargs)
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: I wish there was a way to do the same call in all cases.  Can you think of a way to do that?  It's another place where I wish we didn't have to do a typecheck on the method.  If not, oh well.

File: aiosmtpd/tests/test_handlers.py [hunk 11]
```diff
@@ -1,9 +1,11 @@
+import asyncio
```
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: This should be moved to the line between `sys` and `unittest`

File: aiosmtpd/tests/test_handlers.py [hunk 12]
```diff
@@ -1,9 +1,11 @@
...
 import os
 import sys
 import unittest
 
 from aiosmtpd.controller import Controller
-from aiosmtpd.handlers import Debugging, Mailbox, Message, Proxy, Sink
+from aiosmtpd.handlers import (
+    Debugging, Mailbox, Message, Proxy, Sink, AsyncMessage)
```
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: Please sort this alphabetically.

File: aiosmtpd/tests/test_handlers.py (hunk 6 above)
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: Fair enough!

File: aiosmtpd/tests/test_handlers.py (hunk 11 above)
Comment (sirkonst) [inline aiosmtpd/tests/test_handlers.py]: Why? Now imports sorted in alphabetically, what the point to break it if move `asyncio` under `sys`?

File: aiosmtpd/handlers.py (hunk 1 above)
Comment (sirkonst) [inline aiosmtpd/handlers.py]: Hmm, I don't understand this message, can you suggest how it fix?

File: aiosmtpd/handlers.py (hunk 8 above)
Comment (sirkonst) [inline aiosmtpd/handlers.py]: I moved code from `process_message()` into separated `prepare_message()` for reuse for `AsyncMessage` and avoid copy-past.

File: aiosmtpd/smtp.py (hunk 10 above)
Comment (sirkonst) [inline aiosmtpd/smtp.py]: I think it's impossible.

File: aiosmtpd/tests/test_handlers.py (hunk 11 above)
Comment (sirkonst) [inline aiosmtpd/tests/test_handlers.py]: fixed

**Comment by sirkonst** [comment]
I've updated review and answered for question plz review.

File: aiosmtpd/smtp.py [hunk 13]
```diff
@@ -438,7 +438,13 @@ def smtp_DATA(self, arg):
                 'mail_options': self.mail_options,
                 'rcpt_options': self.rcpt_options,
             }
-        status = self.event_handler.process_message(*args, **kwargs)
+        kwargs.update({'loop': self.loop})
+        if asyncio.iscoroutinefunction(self.event_handler.process_message):
+            kwargs['loop'] = self.loop
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: I think this line is redundant now right?  But actually, I prefer this style over the one in line 441.  Can you just move this up out of the if-else?

File: aiosmtpd/smtp.py (hunk 10 above)
Comment (warsaw) [inline aiosmtpd/smtp.py]: Thanks for the answer.  Do you think we should just make `process_message()` a coroutine in all cases?  That's an API change but I think it would be okay in this case.

File: aiosmtpd/tests/test_handlers.py (hunk 11 above)
Comment (warsaw) [inline aiosmtpd/tests/test_handlers.py]: As long as `tox -e qa` passes, or at least doesn't throw any new warnings over master, I'm happy.

(Note that I'm hoping to improve this as soon as the next version of flake8 comes out.)
